src/iptvsimple/CatchupController.cpp

```diff
@@ -0,0 +1,436 @@
+  {
+    if (m_resetCatchupState)
+    {
+      m_resetCatchupState = false;
+      if (channel.IsCatchupSupported())
+      {
+        m_timeshiftBufferOffset = Settings::GetInstance().GetCatchupDaysInSeconds(); //offset from now to start of catchup window
+        m_timeshiftBufferStartTime = time(nullptr) - Settings::GetInstance().GetCatchupDaysInSeconds(); // now - the window size
+      }
+      else
+      {
+        m_timeshiftBufferOffset = 0;
+        m_timeshiftBufferStartTime = 0;
+      }
+    }
+    else
+    {
+      EpgEntry* currentEpgEntry = GetEPGEntry(channel, m_timeshiftBufferStartTime + m_timeshiftBufferOffset);
+      if (currentEpgEntry)
+        UpdateProgrammeFrom(*currentEpgEntry, channel.GetTvgShift());
+    }
+
+    m_catchupStartTime = m_timeshiftBufferStartTime;
+
+    // TODO: Need a method of updating an inputstream if already running such as web call to stream etc.
+    // this will avoid inputstream restarts which are expensive, may be better placed in client.cpp
+    // this also mean knowing when a stream has stopped
+    SetCatchupInputStreamProperties(true, channel, catchupProperties);
+  }
+}
+
+void CatchupController::ProcessEPGTagForTimeshiftedPlayback(const EPG_TAG& epgTag, Channel& channel, std::map<std::string, std::string>& catchupProperties) // TODO: should be a const channel, create a StreamManager instead
+{
+  TestAndStoreStreamType(channel); // TODO: StreamManager
+
+  if (m_controlsLiveStream)
+  {
+    UpdateProgrammeFrom(epgTag, channel.GetTvgShift());
+    m_catchupStartTime = epgTag.startTime;
+    m_catchupEndTime = epgTag.endTime;
+
+    time_t timeNow = time(0);
+    time_t programmeOffset = timeNow - m_catchupStartTime;
+    time_t timeshiftBufferDuration = std::max(programmeOffset, Settings::GetInstance().GetCatchupDaysInSeconds());
+    m_timeshiftBufferStartTime = timeNow - timeshiftBufferDuration;
+    m_catchupStartTime = m_timeshiftBufferStartTime;
+    m_catchupEndTime = timeNow;
+    m_timeshiftBufferOffset = timeshiftBufferDuration - programmeOffset;
+
+    m_resetCatchupState = false;
+
+    // TODO: Need a method of updating an inputstream if already running such as web call to stream etc.
+    // this will avoid inputstream restarts which are expensive, may be better placed in client.cpp
+    // this also mean knowing when a stream has stopped
+    SetCatchupInputStreamProperties(true, channel, catchupProperties);
+  }
+  else
+  {
+    UpdateProgrammeFrom(epgTag, channel.GetTvgShift());
+    m_catchupStartTime = epgTag.startTime;
+    m_catchupEndTime = epgTag.endTime;
+
+    m_timeshiftBufferStartTime = 0;
+    m_timeshiftBufferOffset = 0;
+
+    m_fromEpgTag = true;
+  }
+}
+
+void CatchupController::ProcessEPGTagForVideoPlayback(const EPG_TAG& epgTag, Channel& channel, std::map<std::string, std::string>& catchupProperties) // TODO: should be a const channel, create a StreamManager instead
+{
+  TestAndStoreStreamType(channel); // TODO: StreamManager
```

Comment:
I do spot so many new TODO's in this PR. I hope the implementation is actually usable? ;-)

Reply:
Many TODO's but still functional!!!!!

Reply:
👍